File: nim/bob/bob.nim
import sequtils
import strutils

func hey*(sentence: string): string =
    let
        trimed = sentence.strip()
        isAlpha = sentence.any(strutils.isAlphaAscii)
        isShouting = strutils.toUpper(trimed) == trimed
        isSilence = trimed == ""
        isQuestion = not isSilence and trimed[^1] == '?'

    if isAlpha and isShouting and isQuestion:
        return "Calm down, I know what I'm doing!"

    if isQuestion:
        return "Sure."

    if isSilence:
        return "Fine. Be that way!"

    if isAlpha and isShouting:
        return "Whoa, chill out!"

    result = "Whatever."

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Text as Text
import Data.Char as Char

containsAlpha :: Text -> Bool
containsAlpha txt = Text.any Char.isAlpha txt

responseFor :: Text -> String
responseFor xs | isEmpty                  = "Fine. Be that way!"
               | isShouting && isQuestion = "Calm down, I know what I'm doing!"
               | isQuestion               = "Sure."
               | isShouting               = "Whoa, chill out!"
               | otherwise                = "Whatever."
               where
                   isQuestion = Text.last (Text.stripEnd xs) == '?'
                   isShouting = containsAlpha xs && Text.toUpper xs == xs
                   isEmpty    = Text.null $ Text.strip xs
